bcgov/cas-registration
# Feat/1608 operation facility form edit

Addresses: [1608](https://github.com/bcgov/cas-registration/issues/1608#issue-2279322344)

### 🚀 Impact:

- Shared ErrorBoundary across apps
   - `bciers/libs/components/src/error/ErrorBoundary.tsx`
   - `bciers/libs/components/src/error/Error.tsx`
   - `bciers/libs/components/src/error/GlobalError.tsx`
- Facility Page href to `add-facility` folder for consistent file naming structure
   - `bciers/apps/administration/app/components/facilities/FacilitiesPage.tsx`
- Facility Form edit 
   - `bciers/apps/administration/app/components/facilities/FacilityForm.tsx`
- Vitests
   - `bciers/apps/administration/tests/components/facilities/FacilityForm.test.tsx`
- Facility endpoint PUT 
   - `bc_obps/registration/api/v1/_facilities/facility_id.py`
- Facility service update
   - `bc_obps/service/facility_service.py`
   - `bc_obps/service/data_access_service/facility_service.py`
- Facility endpoint PUT tests
   - `bc_obps/registration/tests/endpoints/v1/_facilities/test_facility_id.py`


### 📝 Notes
- Hard-code `well_authorization_numbers: [24546, 54321],` in  `bciers/apps/administration/tests/components/facilities/FacilityForm.test.tsx` as  `well_authorization_numbers` will be refactored resulting in refactor required in vitest [see 1888](https://github.com/bcgov/cas-registration/issues/1888)

### 🚧 To Do
[Unauthorized Record Displays in Facilities Grid](https://github.com/bcgov/cas-registration/issues/2019#issue-2435475228)

### 🔬 Local Testing:

1. From terminal command, start the api server:
 ```
cd bc_obps
make reset_db
make run
```  
2.  From terminal command, start the app development server:
 ```
cd client && yarn dev-all
```  
### Test ErrorBoundary component:
1. Navigate to `http://localhost:3000`
2. Click button "Log in with Business BCeID"
3. Sign in to Keycloak using `bc-cas-dev`
4. Navigate to `http://localhost:3000/administration/operations`
![image](https://github.com/user-attachments/assets/11c057db-55a0-4311-bbaa-d2df285f5bd3)
5. Click grid link `Operation 2\View Facilities`
![image](https://github.com/user-attachments/assets/a1ff6e16-bc16-4acd-86b0-858fe96104fb)
6. Click grid link `Facility 2\View Details`
**Expected Results**: Error boundary component displays when nested component throws an error, e.g:
```
throw new Error(
        "We couldn't find your facility information. Please ensure you have been approved for access to this facility.",
      );
```
![image](https://github.com/user-attachments/assets/24c50916-5048-41ea-b31e-376481972c27)


### Test Facility Form Edit:
1. Navigate to `http://localhost:3000`
3. Click button "Log in with Business BCeID"
4. Sign in to Keycloak using `bc-cas-dev`
5. Navigate to `http://localhost:3000/administration/operations`
**Expected Results**   
   - `Operation` displays
![image](https://github.com/user-attachments/assets/11c057db-55a0-4311-bbaa-d2df285f5bd3)
6. Click grid link `Operation 2\View Facilities`
**Expected Results**   
   - `Operation 2 > Facilities` displays
   - 'Add Facilty` button displays
![image](https://github.com/user-attachments/assets/a1ff6e16-bc16-4acd-86b0-858fe96104fb)
6. Click button `Add facility`
**Expected Results**   
   - Facility form in edit mode displays
![image](https://github.com/user-attachments/assets/3e94cbf2-e448-43f7-9720-e2d02b60c357)


### Test Facility Form Edit:
1. Navigate to `http://localhost:3000`
3. Click button "Log in with Business BCeID"
4. Sign in to Keycloak using `bc-cas-dev`
5. Navigate to `http://localhost:3000/administration/operations`
**Expected Results**   
   - `Operation` displays
![image](https://github.com/user-attachments/assets/11c057db-55a0-4311-bbaa-d2df285f5bd3)
6. Click grid link `Operation 2\View Facilities`
**Expected Results**   
   - `Operation 2 > Facilities` displays
![image](https://github.com/user-attachments/assets/a1ff6e16-bc16-4acd-86b0-858fe96104fb)
7. Click grid link `Facility 1\View Details`
**Expected Results**   
   - Record information displays
   - Form fields are read-only
   - `Edit` button displays
   - `Cancel` button displays
![image](https://github.com/user-attachments/assets/ce71be4d-e07a-400e-ad53-1e5c1f2762bb)
8. Click button `Cancel`
**Expected Results**   
   - `Operation 2 > Facilities` displays
![image](https://github.com/user-attachments/assets/94d44100-5d99-495f-bb15-19329bd720fb)
9. Click grid link `Facility 1\View Details`
10. Click button `Edit` 
**Expected Results**   
   - Form fields are editable
   - `Submit` button displays
![image](https://github.com/user-attachments/assets/edf87845-d6c6-4029-9d96-6a2024e69fe0)
11. Add duplicate values in field  `Well Authorization Number(s)`
![image](https://github.com/user-attachments/assets/5ad4c796-5ba5-4875-83d9-8f4940488783)
12. Click button `Submit` 
**Expected Results**
  - Error alert displays: `Well Authorization Number: Well authorization number with this Well authorization number already exists.`
![image](https://github.com/user-attachments/assets/a2f7ad4a-019e-4319-bd20-c4c4332e3595)
13. Add non-duplicate values in `Well Authorization Number(s)` **Note** record the value(s)
14. Complete the other form fields. 
15. Click button `Submit` 
**Expected Results**   
   - Error alert DOES NOT display
   - Sandbox displays: `Your edits were saved successfully`
   - Values entered in the `edit form` reflect in the `read form`
   - `Edit` button displays
   - `Cancel` button displays
![image](https://github.com/user-attachments/assets/eda5be91-df1e-4bb9-81f0-9ebb5152a557)
16. Click button `Cancel`
**Expected Results**   
   - `Operation 2 > Facilities` displays
![image](https://github.com/user-attachments/assets/a1ff6e16-bc16-4acd-86b0-858fe96104fb)
17. Click grid link `Facility 3\View Details`
18. Add values in field  `Well Authorization Number(s)` recorded above (reflected in  `Facility 1\View Details`)
19. Click button `Submit` 
**Expected Results**
  - Error alert displays
![image](https://github.com/user-attachments/assets/648d14d7-65fe-4116-b1c8-22c7c15b933b)


Repeat using `bc-cas-dev-secondary`
**Expected Results**
^ As above.
